File: backend/models/call.py
"""
Call Models

MongoDB schemas for voice/video calling functionality.
Stores call logs, participant info, and call history.
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field


class CallType(str, Enum):
    """Type of call"""
    VOICE = "voice"
    VIDEO = "video"


class CallStatus(str, Enum):
    """Call lifecycle status"""
    RINGING = "ringing"        # Call initiated, waiting for answer
    CONNECTING = "connecting"  # Call accepted, establishing connection
    ACTIVE = "active"          # Call in progress
    ENDED = "ended"            # Call ended normally
    MISSED = "missed"          # Call not answered
    REJECTED = "rejected"      # Call declined by recipient
    FAILED = "failed"          # Technical failure
    BUSY = "busy"              # Recipient on another call


class CallEndReason(str, Enum):
    """Reason why call ended"""
    NORMAL = "normal"              # Normal hang up
    MISSED = "missed"              # Not answered (timeout)
    REJECTED = "rejected"          # Declined by recipient
    BUSY = "busy"                  # Recipient busy
    FAILED = "failed"              # Technical failure
    NETWORK_ERROR = "network_error"
    TIMEOUT = "timeout"
    CANCELLED = "cancelled"        # Caller cancelled before answer


class ParticipantRole(str, Enum):
    """Role in the call"""
    CALLER = "caller"
    CALLEE = "callee"


class CallParticipant(BaseModel):
    """Participant in a call"""
    member_id: str
    member_name: str
    member_avatar: Optional[str] = None
    role: ParticipantRole
    joined_at: Optional[datetime] = None
    left_at: Optional[datetime] = None
    is_muted: bool = False
    is_video_enabled: bool = True
    is_speaker_on: bool = False


class CallCreate(BaseModel):
    """Request to create a new call"""
    call_type: CallType
    caller_id: str
    callee_ids: List[str]  # Support group calls
    community_id: Optional[str] = None
    subgroup_id: Optional[str] = None


class CallResponse(BaseModel):
    """Response after call action"""
    call_id: str
    room_name: str
    livekit_token: str
    livekit_url: str
    call_type: CallType
    participants: List[CallParticipant]


class CallLog(BaseModel):
    """Complete call log entry stored in MongoDB"""
    call_id: str = Field(..., description="Unique call identifier")
    church_id: str = Field(..., description="Church tenant ID")
    room_name: str = Field(..., description="LiveKit room name")
    call_type: CallType
    status: CallStatus = CallStatus.RINGING

    # Participants
    caller: CallParticipant
    callees: List[CallParticipant]

    # Context (optional - for community calls)
    community_id: Optional[str] = None
    community_name: Optional[str] = None
    subgroup_id: Optional[str] = None
    subgroup_name: Optional[str] = None

    # Timestamps
    initiated_at: datetime = Field(default_factory=datetime.utcnow)
    answered_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    duration_seconds: int = 0

    # End reason
    end_reason: Optional[CallEndReason] = None

    # Metadata
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CallHistoryItem(BaseModel):
    """Simplified call history for listing"""
    call_id: str
    call_type: CallType
    status: CallStatus
    caller_id: str
    caller_name: str
    caller_avatar: Optional[str] = None
    callee_ids: List[str]
    callee_names: List[str]
    initiated_at: datetime
    duration_seconds: int
    end_reason: Optional[CallEndReason] = None
    is_incoming: bool  # Relative to the querying user


class CallHistoryResponse(BaseModel):
    """Paginated call history response"""
    calls: List[CallHistoryItem]
    total: int
    page: int
    page_size: int
    has_more: bool


class ActiveCallInfo(BaseModel):
    """Info about an active call"""
    call_id: str
    room_name: str
    call_type: CallType
    status: CallStatus
    participants: List[CallParticipant]
    started_at: datetime
    duration_seconds: int


class CallStatsResponse(BaseModel):
    """Call statistics for analytics"""
    total_calls: int
    total_duration_seconds: int
    voice_calls: int
    video_calls: int
    missed_calls: int
    average_duration_seconds: float

File: mobile/types/call.ts
/**
 * Call Types
 *
 * TypeScript types for voice/video calling functionality.
 * Mirrors backend models for type safety.
 */

// =============================================================================
// Enums
// =============================================================================

export enum CallType {
  VOICE = 'voice',
  VIDEO = 'video',
}

export enum CallStatus {
  RINGING = 'ringing',
  CONNECTING = 'connecting',
  ACTIVE = 'active',
  ENDED = 'ended',
  MISSED = 'missed',
  REJECTED = 'rejected',
  FAILED = 'failed',
  BUSY = 'busy',
}

export enum CallEndReason {
  NORMAL = 'normal',
  MISSED = 'missed',
  REJECTED = 'rejected',
  BUSY = 'busy',
  FAILED = 'failed',
  NETWORK_ERROR = 'network_error',
  TIMEOUT = 'timeout',
  CANCELLED = 'cancelled',
}

export enum ParticipantRole {
  CALLER = 'caller',
  CALLEE = 'callee',
}

// =============================================================================
// Core Types
// =============================================================================

export interface CallParticipant {
  member_id: string;
  member_name: string;
  member_avatar?: string | null;
  role: ParticipantRole;
  joined_at?: string | null;
  left_at?: string | null;
  is_muted: boolean;
  is_video_enabled: boolean;
  is_speaker_on: boolean;
}

export interface CallResponse {
  call_id: string;
  room_name: string;
  livekit_token: string;
  livekit_url: string;
  call_type: CallType;
  participants: CallParticipant[];
}

export interface ActiveCallInfo {
  call_id: string;
  room_name: string;
  call_type: CallType;
  status: CallStatus;
  participants: CallParticipant[];
  started_at: string;
  duration_seconds: number;
}

export interface CallHistoryItem {
  call_id: string;
  call_type: CallType;
  status: CallStatus;
  caller_id: string;
  caller_name: string;
  caller_avatar?: string | null;
  callee_ids: string[];
  callee_names: string[];
  initiated_at: string;
  duration_seconds: number;
  end_reason?: CallEndReason | null;
  is_incoming: boolean;
}

export interface CallHistoryResponse {
  calls: CallHistoryItem[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface CallStatsResponse {
  total_calls: number;
  total_duration_seconds: number;
  voice_calls: number;
  video_calls: number;
  missed_calls: number;
  average_duration_seconds: number;
}

// =============================================================================
// Request Types
// =============================================================================

export interface InitiateCallRequest {
  callee_ids: string[];
  call_type: CallType;
  community_id?: string;
  subgroup_id?: string;
}

export interface RejectCallRequest {
  reason?: 'rejected' | 'busy';
}

export interface EndCallRequest {
  reason?: 'normal' | 'failed' | 'network_error';
}

export interface UpdateParticipantRequest {
  is_muted?: boolean;
  is_video_enabled?: boolean;
  is_speaker_on?: boolean;
}

// =============================================================================
// MQTT Signal Types
// =============================================================================

export enum CallSignalType {
  INVITE = 'call_invite',
  ACCEPT = 'call_accept',
  REJECT = 'call_reject',
  BUSY = 'call_busy',
  END = 'call_end',
  CANCEL = 'call_cancel',
  RINGING = 'call_ringing',
  CONNECTING = 'call_connecting',
  CONNECTED = 'call_connected',
  PARTICIPANT_JOINED = 'participant_joined',
  PARTICIPANT_LEFT = 'participant_left',
  PARTICIPANT_MUTED = 'participant_muted',
  PARTICIPANT_VIDEO = 'participant_video',
  PARTICIPANT_SPEAKER = 'participant_speaker',
  NETWORK_QUALITY = 'network_quality',
}

export interface CallerInfo {
  id: string;
  name: string;
  avatar?: string | null;
}

export interface CallInviteSignal {
  type: CallSignalType.INVITE;
  call_id: string;
  room_name: string;
  call_type: CallType;
  caller: CallerInfo;
  callee_ids: string[];
  community_id?: string | null;
  community_name?: string | null;
  livekit_url: string;
  timestamp: string;
}

export interface CallAcceptSignal {
  type: CallSignalType.ACCEPT;
  call_id: string;
  callee_id: string;
  callee_name: string;
  livekit_token: string;
  timestamp: string;
}

export interface CallRejectSignal {
  type: CallSignalType.REJECT | CallSignalType.BUSY;
  call_id: string;
  callee_id: string;
  callee_name: string;
  reason: string;
  timestamp: string;
}

export interface CallCancelSignal {
  type: CallSignalType.CANCEL;
  call_id: string;
  caller_id: string;
  timestamp: string;
}

export interface CallEndSignal {
  type: CallSignalType.END;
  call_id: string;
  ended_by: string;
  reason: string;
  duration_seconds: number;
  timestamp: string;
}

export interface CallRingingSignal {
  type: CallSignalType.RINGING;
  call_id: string;
  callee_id: string;
  callee_name: string;
  timestamp: string;
}

export interface ParticipantJoinedSignal {
  type: CallSignalType.PARTICIPANT_JOINED;
  call_id: string;
  participant: {
    id: string;
    name: string;
    avatar?: string | null;
  };
  timestamp: string;
}

export interface ParticipantLeftSignal {
  type: CallSignalType.PARTICIPANT_LEFT;
  call_id: string;
  participant_id: string;
  participant_name: string;
  timestamp: string;
}

export interface ParticipantMutedSignal {
  type: CallSignalType.PARTICIPANT_MUTED;
  call_id: string;
  participant_id: string;
  is_muted: boolean;
  timestamp: string;
}

export interface ParticipantVideoSignal {
  type: CallSignalType.PARTICIPANT_VIDEO;
  call_id: string;
  participant_id: string;
  is_video_enabled: boolean;
  timestamp: string;
}

export interface ParticipantSpeakerSignal {
  type: CallSignalType.PARTICIPANT_SPEAKER;
  call_id: string;
  participant_id: string;
  is_speaker_on: boolean;
  timestamp: string;
}

export interface NetworkQualitySignal {
  type: CallSignalType.NETWORK_QUALITY;
  call_id: string;
  participant_id: string;
  quality: 'excellent' | 'good' | 'fair' | 'poor';
  latency_ms: number;
  packet_loss_percent: number;
  timestamp: string;
}

export type CallSignal =
  | CallInviteSignal
  | CallAcceptSignal
  | CallRejectSignal
  | CallCancelSignal
  | CallEndSignal
  | CallRingingSignal
  | ParticipantJoinedSignal
  | ParticipantLeftSignal
  | ParticipantMutedSignal
  | ParticipantVideoSignal
  | ParticipantSpeakerSignal
  | NetworkQualitySignal;

// =============================================================================
// UI State Types
// =============================================================================

export type CallUIState =
  | 'idle'
  | 'outgoing'      // Initiating call, waiting for answer
  | 'incoming'      // Receiving call invitation
  | 'connecting'    // Call accepted, establishing connection
  | 'active'        // Call in progress
  | 'reconnecting'  // Reconnecting after network issue
  | 'ended';        // Call ended

export interface CallDuration {
  hours: number;
  minutes: number;
  seconds: number;
  formatted: string;  // "00:00" or "1:23:45"
}

export interface NetworkQuality {
  level: 'excellent' | 'good' | 'fair' | 'poor' | 'unknown';
  latencyMs?: number;
  packetLossPercent?: number;
}

// =============================================================================
// LiveKit Types (for React Native SDK)
// =============================================================================

export interface LiveKitConfig {
  url: string;
  token: string;
  roomName: string;
}

export interface LocalParticipantState {
  isMuted: boolean;
  isVideoEnabled: boolean;
  isSpeakerOn: boolean;
  isScreenSharing: boolean;
}

export interface RemoteParticipantState {
  participantId: string;
  name: string;
  avatar?: string;
  isMuted: boolean;
  isVideoEnabled: boolean;
  isSpeaking: boolean;
  networkQuality: NetworkQuality;
}

// =============================================================================
// Call Store State Types
// =============================================================================

export interface CallStoreState {
  // Current call state
  uiState: CallUIState;
  currentCall: ActiveCallInfo | null;
  incomingCall: CallInviteSignal | null;

  // LiveKit connection
  livekitConfig: LiveKitConfig | null;
  isConnected: boolean;

  // Local participant
  localParticipant: LocalParticipantState;

  // Remote participants
  remoteParticipants: Map<string, RemoteParticipantState>;

  // Call timing
  callStartTime: Date | null;
  callDuration: CallDuration;

  // Network
  networkQuality: NetworkQuality;

  // Error handling
  error: string | null;
}

export interface CallStoreActions {
  // Call lifecycle
  initiateCall: (callee_ids: string[], call_type: CallType, community_id?: string) => Promise<void>;
  acceptCall: (call_id: string) => Promise<void>;
  rejectCall: (call_id: string, reason?: 'rejected' | 'busy') => Promise<void>;
  cancelCall: () => Promise<void>;
  endCall: (reason?: 'normal' | 'failed' | 'network_error') => Promise<void>;

  // Incoming call handling
  handleIncomingCall: (signal: CallInviteSignal) => void;
  handleCallCancelled: (signal: CallCancelSignal) => void;

  // Signal handling
  handleCallAccepted: (signal: CallAcceptSignal) => void;
  handleCallRejected: (signal: CallRejectSignal) => void;
  handleCallEnded: (signal: CallEndSignal) => void;
  handleParticipantJoined: (signal: ParticipantJoinedSignal) => void;
  handleParticipantLeft: (signal: ParticipantLeftSignal) => void;
  handleParticipantMuted: (signal: ParticipantMutedSignal) => void;
  handleParticipantVideo: (signal: ParticipantVideoSignal) => void;

  // Local controls
  toggleMute: () => void;
  toggleVideo: () => void;
  toggleSpeaker: () => void;
  switchCamera: () => void;

  // Duration timer
  startDurationTimer: () => void;
  stopDurationTimer: () => void;

  // Reset
  reset: () => void;
}

export type CallStore = CallStoreState & CallStoreActions;
